server/grpc_service.go

```diff
@@ -0,0 +1,534 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/pdpb2"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+// notLeaderError is returned when current server is not the leader and not possible to process request.
+// TODO: work as proxy.
+var notLeaderError = grpc.Errorf(codes.Unavailable, "not leader")
+
+// GetPDMembers implements gRPC PDServer.
+func (s *Server) GetPDMembers(context.Context, *pdpb2.GetPDMembersRequest) (*pdpb2.GetPDMembersResponse, error) {
+	members, err := GetPDMembers(s.GetClient())
+	if err != nil {
+		return nil, grpc.Errorf(codes.Unknown, err.Error())
+	}
+	members2 := make([]*pdpb2.Member, len(members))
+	for i := range members {
+		members2[i] = &pdpb2.Member{
+			Name:       members[i].GetName(),
+			MemberId:   s.clusterID,
+			PeerUrls:   members[i].PeerUrls,
+			ClientUrls: members[i].ClientUrls,
+		}
+	}
+	return &pdpb2.GetPDMembersResponse{
+		Header:  s.header(),
+		Members: members2,
+	}, nil
+}
+
+// Tso implements gRPC PDServer.
+func (s *Server) Tso(stream pdpb2.PD_TsoServer) error {
+	for {
+		request, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if !s.IsLeader() {
+			return notLeaderError
+		}
+		if err := s.validateHeader(request.GetHeader()); err != nil {
```

Comment:
I have also considered it, but seems we need etcd to export its internal grpc server instance.